### PR TITLE
fix(bot): disable Kodiak on GitHub API merge internal server error

### LINKED ISSUE #397: It's unsafe to retry merge when GitHub's pull request merge API returns a 500 error
Kodiak has merged the same PR multiple times. It wasn't until I manually clicked the merge button that it stopped.

https://github.com/zeit/now/commits/master

https://github.com/zeit/now/pull/4412

image

### PR DESCRIPTION
**Pull request by chdsbd**
This change disables Kodiak merging a pull request when it encounters an internal server error from the GitHub merge api. In this event, Kodiak adds a label `disable_kodiak_label` and leaves a comment describing the issue. Kodiak will not update, approve, comment on, label, or merge a pull request with the `disable_kodiak_label`. It will still update status checks. The user must remove this label to re-enable Kodiak merging the pull request.

fixes #397

## background
It is dangerous to retry merging when GitHub returns an internal server error (HTTP status code 500) because the merge can partially succeed in a way that will allow Kodiak to erroneously merge in a pull request multiple times, as we've seen in #397.

The GitHub pull request merge API is _not always_ idempotent (https://developer.github.com/v3/pulls/#merge-a-pull-request-merge-button). Normally if you try to merge an already merged pull request using this API you'll get a `405 Method not allowed` error, as documented in the GitHub API docs. However in an error state, when the GitHub API returns a 500 HTTP error, the pull request can be partially merged, as shown in #397.

It appears the git operations of merging the pull request branch into master can succeed but not affect the pull request state itself. In this failure scenario the pull request will still be in an OPEN state even though the underlying branch was merged successfully into the target branch.

## further discussion
I realized implementing this change that GitHub API calls to add pull request labels could also fail. In this case Kodiak would still erroneously merge a pull request.

I think the most robust solution to this GitHub API reliability problem would be to store some information in Redis to disable Kodiak. But then there's the problem of how to surface that data in the GitHub UI. If we can't trust the GitHub API, how can we notify a user that Kodiak is disabled on their pull request? 🙍‍♂️ 

## screenshots
image

### CODE REVIEW

File: bot/kodiak/evaluation.py [hunk 1]
```diff
@@ -750,11 +757,44 @@ async def set_status(msg: str, markdown_content: Optional[str] = None) -> None:
...
+        try:
+            await api.merge(
+                merge_method=merge_args.merge_method,
+                commit_title=merge_args.commit_title,
+                commit_message=merge_args.commit_message,
+            )
+        # if we encounter an internal server error (status code 500), it is
+        # _not_ safe to retry. Instead we mark the pull request as unmergable
+        # and require a user to re-enable Kodiak on the pull request.
+        except GitHubApiInternalServerError:
```
Comment (chdsbd) [inline bot/kodiak/evaluation.py:769]: I know we have some retry logic for API exceptions that runs outside of `mergeable()`, but I think this is the proper place to handle this error. I think we should also try to move that retry logic in to the function as well.

We can easily provide thorough test coverage of any logic inside of `mergable()`.

File: bot/kodiak/pull_request.py [hunk 2]
```diff
@@ -283,13 +288,30 @@ async def merge(
...
+        self.log.info("add_label", label=label)
+        async with self.client(
+            installation_id=self.install, owner=self.owner, repo=self.repo
+        ) as api_client:
+            res = await api_client.add_label(label, pull_number=self.number)
+            try:
+                res.raise_for_status()
+            except HTTPError:
+                self.log.exception("failed to delete label", label=label, res=res)
+                raise ApiCallException("add label")
```
Comment (chdsbd) [inline bot/kodiak/pull_request.py:313]: I realized while adding this endpoint that if we encounter an exception trying to add a label to a pull request we will not be able to disable the bot.

I'm not sure what to do about this.

File: bot/kodiak/config.py [hunk 3]
```diff
@@ -124,6 +124,11 @@ class V1(BaseModel):
     merge: Merge = Merge()
     update: Update = Update()
     approve: Approve = Approve()
+    # label to apply to pull request to disable Kodiak from taking any action
```
Comment (sbdchd) [inline bot/kodiak/config.py:127]:
```suggestion
    # when added to a Pull Request Kodiak will be prevented from taking any action
```

**Comment by chdsbd** [comment]
This change has been deployed.

**Pull request merged**